File: Python/indexSin.py
"""IndexSin lookup table for sin/cos with index-frequency input (resolution
pi/size, size = 1<<order). Provides Quart/Full/Fixed SinSource variants used
by FFT and regression-based sin generation to minimise float error.
"""
import enum
import math
import os

import varDbl

# Resolve OUTDIR (project root) from this file's location, independent of CWD.
_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
OUTDIR = os.path.dirname(_THIS_DIR)


class SinSource (enum.StrEnum):
    Quart = 'Quart',
    Lib = 'Lib',
    Prec = 'Prec'


class IndexSin:
    '''
    A sin() with index frequence as input, with resolution PI /(1 << "order)
    '''
    MIN_ORDER = 1
    MAX_ORDER = 18

    _size = 1 << MAX_ORDER
    _half = 1 << (MAX_ORDER - 1)

    _sSinQuart = None
    _sSinPrec = None

    __slots__ = ['_order', '_sinSource', '_sSin', '_sCos', '_order']

    @staticmethod
    def validateOrder(order:int):
        if not (IndexSin.MIN_ORDER <= order <= IndexSin.MAX_ORDER):
            raise RuntimeError(f'order={order} is not in the range of [{IndexSin.MIN_ORDER}, {IndexSin.MAX_ORDER}]')

    @staticmethod
    def validateSize(size:int) -> int:
        for order in range(IndexSin.MIN_ORDER, IndexSin.MAX_ORDER + 1):
            if (1 << order) == size:
                return order
        else:
            raise RuntimeError(f'Invalid input array size {size} which is not 2^{order}')


    def __init__(self, sinSource = SinSource.Quart,
                 sCosSin:tuple[varDbl.VarDbl]=None) -> None:                
        self._sinSource = sinSource
        match sinSource:
            case SinSource.Prec:
                if not IndexSin._sSinPrec:
                    IndexSin._sSinPrec = IndexSin.read(SinSource.Prec)
                self._order = IndexSin.MAX_ORDER
                self._sSin = IndexSin._sSinPrec
                self._sCos =  None
            case SinSource.Quart:
                if not IndexSin._sSinQuart:
                    quart = IndexSin._half >> 1
                    sSin = [math.sin(math.pi * i /IndexSin._size) for i in range(quart)] + \
                        [math.cos(math.pi * (quart - i) / IndexSin._size) for i in range(quart + 1)]
                    IndexSin._sSinQuart = tuple([varDbl.VarDbl(v) for v in sSin])
                self._order = IndexSin.MAX_ORDER
                self._sSin = IndexSin._sSinQuart
                self._sCos =  None
            case SinSource.Lib:
                self._order = None
                self._sSin = None
                self._sCos =  None
            case _:
                raise RuntimeError(f'sinSource={sinSource}')

    @staticmethod
    def header():
        return 'Index\tX\tValue\tUncertainty\n'

    def dump(self, order:int=MAX_ORDER) -> None:
        IndexSin.validateOrder(order)
        with open(f'{OUTDIR}/Python/Output/IndexSin_{self._sinSource}_{order}.txt', 'w') as f:
            f.write(IndexSin.header())
            size = 1 << order
            for i in range(size + 1):
                v = self.sin(i, order)
                f.write(f'{i}\t{i/size:.20e}\t{v.value():.20e}\t{v.uncertainty():.20e}\n')

    @staticmethod
    def read(sinSource:SinSource, order:int=MAX_ORDER, errorFold:float=1e6) -> list[varDbl.VarDbl]:
        '''
        {errorFold} indicates the difference between C++ long double and python double
        '''
        IndexSin.validateOrder(order)
        size = 1 << order
        dumpPath = os.path.join(OUTDIR, 'Cpp', 'Output', f'IndexSin_{sinSource}_{order}.txt')
        sSin = []
        with open(dumpPath) as f:
            hdr = next(f)
            if hdr != IndexSin.header():
                raise RuntimeError(f'Invalid header in file {dumpPath}: {hdr} vs {IndexSin.header()}')
            for ln, line in enumerate(f):
                sWord = line.split('\t')
                if ln != int(sWord[0]):
                    raise RuntimeError(f'Invalid index {sWord[0]} vs {ln} in {dumpPath}: {line}')
                phase = ln / size
                if float(sWord[1]) != phase:
                    raise RuntimeError(f'Invalid phase {sWord[1]} vs {phase} for index {ln} in {dumpPath}: {line}')
                sin = math.sin(math.pi * phase)
                val, unc = map(float, sWord[2:])
                if unc < 0:
                    raise RuntimeError(f'Invalid uncertainty {sWord[3]} for index {ln} in {dumpPath}: {line}')
                if abs(val - sin) > unc * errorFold:
                    if unc > 0:
                        raise RuntimeError(f'Invalid value {sWord[2]} and uncertainty {sWord[3]} for sin={sin} diffFold={(val - sin)/unc} at index {ln} in {dumpPath}: {line}')
                sSin.append(varDbl.VarDbl(val, unc))
            if len(sSin) < (size >> 1) + 1:
                raise RuntimeError(f'Invalid sin count {len(sSin)} < {(size >> 1) + 1} in {dumpPath}: {line}')
            return sSin

    @property  
    def sinSource(self) -> int:
        return self._sinSource
        
    def get_index(self, freq:int, order:int) ->int:
        '''
        get index into _sSin, with -index means -sin
        '''
        size = 1 << order
        div = freq // size
        rem = freq % size
        if ((self.sinSource == SinSource.Prec) or (self.sinSource == SinSource.Quart)) \
            and (rem > (size >> 1)):
            rem = size - rem
        if div & 1:
            return - rem
        else:       
            return rem


    def sin(self, freq:int, order:int) -> varDbl.VarDbl:
        if self.sinSource == SinSource.Lib:
            v = math.sin(math.pi * freq / (1 << order))
            return varDbl.VarDbl(v)
        else:
            IndexSin.validateOrder(order)
            idx = self.get_index(freq << (IndexSin.MAX_ORDER - order), IndexSin.MAX_ORDER)
            if idx >= 0:
                return self._sSin[idx]
            else:
                return -self._sSin[-idx]

    def cos(self, freq:int, order:int) -> varDbl.VarDbl:
        if self.sinSource == SinSource.Lib:
            v = math.cos(math.pi * freq / (1 << order))
            return varDbl.VarDbl(v)
        else:
            IndexSin.validateOrder(order)
            return self.sin(freq + (1 << (order - 1)), order)
      
        

File: Java/src/Func/IndexSin.java
package Func;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

import Type.InitException;
import Type.VarDbl;


enum SinSource {
    Prec,
    Quart,
    Lib
}


/*
 * A cached sine function with resolution down to PI /2^order().
 * It has integer input to sin(), cos() and tan() to avoid rounding errors.
 */
public class IndexSin {

    /*
     * The maximual order of FFT calculation.
     * The size of FFT is limited to MAX_QUARD_ORDER
     */

    static public final int MIN_ORDER = 1;
    static public final int MAX_ORDER = 18;

    private static final VarDbl[] _sPrecSin = new VarDbl[(1 << (MAX_ORDER - 1)) + 1];

    private static final VarDbl[] _sQuartSin = new VarDbl[(1 << (MAX_ORDER - 1)) + 1];
    static {
        final int size = 1 << MAX_ORDER;
        final int half = size >> 1;
        final int quart = half >> 1;
        _sQuartSin[0] = new VarDbl(0);
        _sQuartSin[half] = new VarDbl(1);
        try {
            _sQuartSin[quart] = new VarDbl(Math.sin(Math.PI/4));
        } catch(InitException ex) {
            throw new RuntimeException(
                String.format("The index %d/%d for IndexSin has exception %s", size, MAX_ORDER, ex));
        }
        final double f = Math.PI/size;
        for (int i = 1; i < quart; ++i) {
            try {
                _sQuartSin[i] = new VarDbl(Math.sin(f*i));
                _sQuartSin[half - i] = new VarDbl(Math.cos(f*i));
            } catch(InitException ex) {
                throw new RuntimeException(
                    String.format("The index %d/%d for IndexSin has exception %s", i, MAX_ORDER, ex));
            }
        }
    }


    public final SinSource sinSource;
 
    public IndexSin(SinSource sinSource) {
        if (sinSource == SinSource.Prec) {
            if (_sPrecSin[0] == null)
                readPrec();
        } else if (sinSource != SinSource.Quart && sinSource != SinSource.Lib) {
            throw new IllegalArgumentException(
                String.format("Unknown SinSource %s for IndexSin", sinSource));
        }
        this.sinSource = sinSource;
    }

    void readPrec() {
        final String dumpPath = "./Cpp/Output/IndexSin_Prec_18.txt";
        try (final BufferedReader br = new BufferedReader(new FileReader(dumpPath))) {
            String line = br.readLine();
            if (! line.equals("Index\tX\tValue\tUncertainty"))
                throw new IllegalArgumentException(String.format("Invalid header in file %s: %s", dumpPath, line));
            for (int i = 0; i < _sPrecSin.length; ++i) {
                line = br.readLine();
                if (line == null)
                    throw new IllegalArgumentException(String.format("Invalid %d null line in file %s", i, dumpPath));
                final String[] sWord = line.trim().split("\t");
                if (sWord.length != 4)
                    throw new IllegalArgumentException(String.format("Invalid %d line with %d colums in file %s: %s", i, sWord.length, dumpPath, line));
                try {
                    final int n = Integer.parseInt(sWord[0]);
                    if (n != i)
                        throw new IllegalArgumentException(String.format("Invalid %d line index %d in file %s: %s", i, n, dumpPath, line));
                } catch (NumberFormatException e) {
                    throw new IllegalArgumentException(String.format("Invalid %d line at index count in file %s: %s", i, dumpPath, line));
                }
                try {
                    final double x = Double.parseDouble(sWord[1]);
                    final double rad = ((double) i) / (1 << MAX_ORDER);
                    if (Math.abs(x - rad) > Math.ulp(1))
                        throw new IllegalArgumentException(String.format("Invalid %d line x %e vs %e in file %s: %s", i, x, rad, dumpPath, line));
                } catch (NumberFormatException e) {
                    throw new IllegalArgumentException(String.format("Invalid %d line at index count in file %s: %s", i, dumpPath, line));
                }
                final double val, unc;
                try {
                    val = Double.parseDouble(sWord[2]);
                } catch (NumberFormatException e) {
                    throw new IllegalArgumentException(String.format("Invalid %d line at value '%s' in file %s: %s", i, sWord[2], dumpPath, line));
                }
                try {
                    unc = Double.parseDouble(sWord[3]);
                    if ((unc < 0) || (unc > 2.*Math.ulp(val)))
                       throw new IllegalArgumentException(String.format("Invalid %d line uncertainty %e in file %s: %s", i, unc, dumpPath, line));
                } catch (NumberFormatException e) {
                    throw new IllegalArgumentException(String.format("Invalid %d line at value '%s' in file %s: %s", i, sWord[2], dumpPath, line));
                }
                try {
                    _sPrecSin[i] = new VarDbl(val, unc);
                } catch (InitException e) {
                    throw new IllegalArgumentException(String.format("Invalid %d line of VarDbl(%e, %e) in file %s: %s", i, val, unc, dumpPath, line));
                }
            }
        } catch (IOException e) {   
            throw new IllegalArgumentException(String.format("Invalid Prec file '%s': %s", dumpPath, e.getMessage()));
        } 
    }

    int get_index(long freq, int order) {
        if (order < MIN_ORDER) {
             throw new IllegalArgumentException(
                String.format("The order %d < %d for IndexSin", order, MIN_ORDER));
        }
        if (MAX_ORDER < order) {
            throw new IllegalArgumentException(
                String.format("The order %d > %d for IndexSin", order, MAX_ORDER));
        }
        final int half = 1 << (order - 1);
        boolean neg = freq < 0;
        long div = Math.abs(freq) / half;
        long rem = Math.abs(freq) % half;
        if ((div & 1) == 1) {
            rem = half - rem;
        }
        if ((div & 2) == 2) {
            neg = !neg;
        }
        return (int) (neg? -rem : rem);
    }
    
    public VarDbl sin(long freq, int order) {
        if (sinSource == SinSource.Quart) {
            final int idx = this.get_index(freq, order);
            final VarDbl ret = new VarDbl(_sQuartSin[Math.abs(idx) << (MAX_ORDER - order)]);
            return (idx >= 0)? ret : ret.negate();
        } else if (sinSource == SinSource.Prec) {
            final int idx = this.get_index(freq, order);
            final VarDbl ret = new VarDbl(_sPrecSin[Math.abs(idx) << (MAX_ORDER - order)]);
            return (idx >= 0)? ret : ret.negate();
        } else if (sinSource == SinSource.Lib) {
            try {
                double ret = Math.sin(Math.PI / (1 << order) * freq);
                return new VarDbl(ret);
            } catch (InitException e) {
                throw new RuntimeException(String.format(
                    "The freq %d/%d for IndexSin.sin()", freq, order));
            }
        } else {
            throw new IllegalArgumentException(
                String.format("Unknown SinSource %s for IndexSin.sin()", sinSource));
        }
    }

    public VarDbl cos(long freq, int order) {
        return sin(freq + (1 << (order - 1)), order);
    }

    public boolean dump(final String dumpPath, int order) {
        try (FileWriter writer = new FileWriter(dumpPath)) {
            writer.write("Index\tX\tValue\tUncertainty");
            final double f = 1.0 / (1 << order);
            for (int i = 0; i < _sQuartSin.length; ++i) {
                final VarDbl sin = sin(i, order);
                writer.write(String.format("\n%d\t%.20e\t%.20e\t%.20e", 
                             i, i *f, sin.value(), sin.uncertainty()));
            }
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }            
        return true;
    }
}
